File: Algorithms/712.Minimum-ASCII-Delete-Sum-for-Two-Strings/minimum_ascii_delete_sum_for_two_strings.go
func minimumDeleteSum(s1 string, s2 string) int {
	len1, len2 := len(s1), len(s2)
	dp := make([]int, len2+1)
	for i := range dp {
		dp[i] = 0
	}

	for i := 1; i < len2+1; i++ {
		dp[i] = dp[i-1] + int(s2[i-1])
	}

	for i := 1; i < len1+1; i++ {
		temp_j_one := dp[0]
		dp[0] += int(s1[i-1])
		for j := 1; j < len2+1; j++ {
			temp_j := dp[j]
			if s1[i-1] == s2[j-1] {
				dp[j] = temp_j_one
			} else {
				dp[j] = min(temp_j+int(s1[i-1]), dp[j-1]+int(s2[j-1]))
			}
			temp_j_one = temp_j
		}
	}
	return dp[len2]
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: Algorithms/712.Minimum-ASCII-Delete-Sum-for-Two-Strings/minimum-ascii-delete-sum-for-two-strings.py
# -*- coding: utf-8 -*-

class Solution:
    def minimumDeleteSum(self, s1, s2):
        """
        :type s1: str
        :type s2: str
        :rtype: int
        """
        len1, len2 = len(s1), len(s2)
        dp = [[0] * (len2 + 1) for _ in range (len1 + 1)]
        for i in range(len1):
            dp[i + 1][0] = dp[i][0] + ord(s1[i])
        for j in range(len2):
            dp[0][j + 1] = dp[0][j] + ord(s2[j])
        for i in range(1, len1 + 1):
            for j in range(1, len2 + 1):
                if s1[i - 1] == s2[j - 1]:
                    dp[i][j] = dp[i - 1][j - 1]
                else:
                    dp[i][j] = min(dp[i - 1][j] + ord(s1[i - 1]), dp[i][j - 1] + ord(s2[j - 1]))
        return dp[-1][-1]
        
